_posts/2021-01-27-#20833.md

```diff
@@ -0,0 +1,164 @@
+  Importantly, if a transaction is invalid, `MemPoolAccept` is designed such that
+  it fails as quickly as possible to mitigate DoS risks.
+
+- Each `MemPoolAccept` instance creates a Coins View from the current chain tip
+  and mempool.  `CCoinsView` is a class with a hierarchical structure (see
+  diagram below).
+
+```
+             +--------------------+
+             |                    |
+             |     CCoinsView     |
+             |                    |
+             +----------+---------+
+                        ^
+                        |
+             +----------+---------+
+             |                    |
+             |  CCoinsViewBacked  |
+             |                    |
+             +------+------+------+
+                    ^      ^
+           +--------+      +--------+
+           |                        |
++----------+---------+  +-----------+---------+
+|                    |  |                     |
+|  CCoinsViewCache   |  |  CCoinsViewMemPool  |
+|                    |  |                     |
++--------------------+  +---------------------+
+```
+
+- A `CCoinsViewBacked` points to another `CCoinsView` as its "backend." When
+  `CCoinsViewBacked.GetCoin()` cannot find a coin in its own data structures,
+  it calls `backend.GetCoin()` to continue searching. Thanks to the magic of
+  polymorphism, `CCoinsViewBacked` Coins Views can be "layered" to allow
+  searching for coins in multiple locations without changing the underlying chain
+  state.
+
+- Since the UTXO set is quite large and stored on disk, Bitcoin Core uses a
+  cache to improve validation performance.  To prevent cache thrashing caused
+  by an attacker broadcasting many invalid transactions spending a wide variety
+  of UTXOs (a cheap DoS attack), we keep track of which Coins are brought in from
+  disk and uncache them if the transaction turns out to be invalid.
+
+- Validation roughly consists of `PreChecks`, `PolicyScriptChecks`,
+  `ConsensusScriptChecks`, and `Finalize`.  We've touched on these functions in
+  previous review clubs, e.g.  [#19339](https://bitcoincore.reviews/19339).  This
+  [Stack Exchange
+  post](https://bitcoin.stackexchange.com/questions/100317/what-is-the-difference-between-policy-and-consensus-when-it-comes-to-a-bitcoin-c)
+  about consensus vs policy rules may also be helpful.
+
+- In an effort to not make any API-breaking changes, #20833 returns a `results`
+  array with one result per transaction in the same order they were passed in
+  or the first failure result.
+
+## Questions
+
+1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
+   NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?
+
+2. What are some use cases for enabling packages through `testmempoolaccept`?
+
+3. What are some limitations of this implementation of package validation?
+
+4. From a code niceness standpoint, what is the benefit of moving the
```

Comment:
'niceness' :laughing: 

Maybe s/niceness/quality/ ?

Reply:
Would you believe me if I said I spent like 10 minutes looking for a better word 😅